Differentiation between standards and decision records (#352) (#441)
SovereignCloudStack/standards: Standards/scs-XXXX-vN-decision-record-template.md

Author: To start a discussion on this (and take the suggestion of @markus-hentsch from #352):

Is a versioning of Decision Records necessary?

I can understand both cases. On one hand, the Standards and DRs of SCS are designed a bit like RFC standards, meaning that certain standards supersed previous ones, which is mostly indicated by the version number of a standard. This is good to keep track of changes and explain certain "historic" decisions and maybe old setups that still use older standards. The same would be true for Decision Records.
On the other hand we must acknowledge that Decision Records are not really binding and therefore are not  required to keep track of their changes.

Reviewer: IMO, we do not need versioning in DR. Most of them will not be touched after they where written. Is suggest to create e new DR, if the old one must be reconsidered, I suggest to open a new DR referencing the old one.

---
title: _Descriptive title_

> type: Standard  # | Procedural

Reviewer 2: You seem to have mixed this up with the standard template.

> status: Draft

Reviewer 3: should this always be `Draft`?

Author: Nope, I just didn't change it, since the previous Standard document also had only `Draft` there, probably to suggest that a completely new document can't be stable right from the start.

Reviewer 2: Yes. Initially, it may only be a draft.

Reviewer 3: This information should be reflected somewhere in the document though, no?

Reviewer 2: These templates are meant to be used by people who have read and understood scs-0001-v1.

track: Global  # | IaaS | Ops | KaaS | IAM
author: Author <Mail or Github handle>
date: DD-MM-YYYY
---

<!---
This is a template striving to provide a starting point for
creating a standard or decision record adhering to scs-0001.
Replace at least all text which is _italic_.
See https://github.com/SovereignCloudStack/standards/blob/main/Standards/scs-0001-v1-sovereign-cloud-standards.md
--->

## Introduction

Introduction

## Terminology

> Terminology

Reviewer 3: This needs - imho - more explanation or an example.

Author: I had something in mind like its done in some RFC documents e.g. [HTTP](https://www.rfc-editor.org/rfc/rfc2616) with their section `Terminology`
But how exactly this section is built up is up for discussion. I would like to do it like the mentioned `RFC` section, but I'm also on board with the previously used table structure.

Author: Updated the section to contain an example for a terminology.

---

> title: _Descriptive title_

Reviewer 2: I think @markus-hentsch somewhere remarked that there were no parts in italics/emphasized, but this is a counter example.

type: Decision Record
status: Draft
track: Global  # | IaaS | Ops | KaaS | IAM

> author: Author <Mail or Github handle>
> date: DD-MM-YYYY

Reviewer 2: ```suggestion
```

Author: Changed in 6fb17a19fc38f70a0dda6c5cc9689ff28d0b31e4

---

<!---
This is a template striving to provide a starting point for

> creating a standard or decision record adhering to scs-0001.

Reviewer 4: ```suggestion
creating a decision record adhering to scs-0001.
```

This is only about DRs by now, no?

Author: _https://github.com/SovereignCloudStack/standards/pull/441/commits/333f64e6cce05cceff6e49087cb595e310ab882f_

> Replace at least all text which is _italic_.

Reviewer 4: ```suggestion
Replace at least all text in the sections not marked as OPTIONAL.
```

(there's no italic text in this template)

Author: https://github.com/SovereignCloudStack/standards/pull/441/commits/333f64e6cce05cceff6e49087cb595e310ab882f

See https://github.com/SovereignCloudStack/standards/blob/main/Standards/scs-0001-v1-sovereign-cloud-standards.md
--->

> ## Introduction

Reviewer 4: Do we actually need an introduction section for a DR? I can see it making sense for a standard but if I imagine the situation of needing to write a DR in my head, I would want the process to be short and concise. I can't imagine anything I'd want to write here when documenting a decision since I'd put any surrounding information into the "Context" section anyways.

Author: https://github.com/SovereignCloudStack/standards/pull/441/commits/1795a118b1bbc30bab68d2b64c36f30d267be099

Reviewer 4: As discussed in the SIG Standardization/Certification today, we will reintroduce this section as "Abstract" aiming to briefly introduce what the decision (record) is about in one or two sentences on a high level. "Context" would then be used to more deeply describe technical details.

I'm fine with this separation. Resolving this thread.

Introduction

## Terminology

Example (abbr. Ex)
  This is the description for an example terminology.

## Context

> What is the issue that we're seeing that is motivating this decision or change?

Reviewer 3: Should this not be called `Motivation` like mentioned [here](https://github.com/SovereignCloudStack/standards/blob/main/Standards/scs-0001-v1-sovereign-cloud-standards.md#sections)

Author: Not necessarily, according to @mbuechse we would need to adapt the standard for these changes. This was a step that still needs to be done (thats why the PR is still in Draft)

Author: Adapted the standard in a Version 2 in order to reflect the document templates `Standard` and `Decision Record`.

> ---

Author: Another thing that needs to be discussed is the structure of this DR template.

I tried to derive it from the ADR templates found here https://github.com/joelparkerhenderson/architecture-decision-record/tree/main/locales/en/templates, but there are many different ideas on the web how a DR should be handled. Some examples:
https://docs.aws.amazon.com/prescriptive-guidance/latest/architectural-decision-records/adr-process.html
https://www.ozimmer.ch/practices/2022/11/22/MADRTemplatePrimer.html
http://www.iso-architecture.org/42010/templates/42010-ad-template.pdf

I wouldn't put myself out there to say what is right or wrong. The gist I collected from reading multiple articles is that the following information should be included:
* What issue is the DR deciding on and why?
* What options were considered and why?
* What is the final decision?
* How does this affect the product/current status (positive and negative)?
* (Who was involved in the decision and when was it done?)

I'm up for a discussion on this, if someone else has a stronger opinion here, but I also want to say that this whole topic is probably very much opinionated. I only I would write a DR, I would just include the actual decision and the reasoning behind it, which would also include an explanation why certain options where not picked. This is enough IMO, since certain topics can become very hard to unravel and someone could always ask "Why didn't you do X?". Time is always a factor and also with DRs (or standards), you can only look at a certain number of options before making a decision, which is why an explanation for not picking some option isn't really always that useful. A good explanation on why some option was chosen makes more sense here.

Reviewer 3: After further thinking about this, I would maybe only add one additional optional bullet point to the above list:

* What options were intentionally not considered and why?

This should only be used when it makes sense to do so, e.g. "we didn't consider option X because the time to investigate it was $foobar, also other $reason".

Maybe this doesn't need to be spelled out, but I often find myself reading documentation and wondering why option X was never considered and when asking people it turns out it was not considered for good reasons - so someone thought about that already - but this was never spelled out anywhere. I recognize though that of course you can't document every option you didn't consider, so this only makes sense for very prominent options, that come to mind immediately.
